File: work/arc060_d.cpp
#include<iostream>
#include<cmath>
using namespace std;

typedef long long ll;


ll f(ll b, ll n){
  ll ret = 0;
  while(n>=b){
    ret += n%b;
    n = n/b;
  }
  return ret+n;
}

int main(){
  ll n,s;
  cin>>n>>s;
  if(s==n){
    cout<<n+1<<endl;
    return 0;
  }

  ll b=2;
  while(b*b<=n){
    if(f(b,n)==s){
      cout<<b<<endl;
      return 0;
    }
    b += 1;
  }

  ll p = ll(sqrt(n))+2;
  while(p>0){
    ll q = s - p;
    ll b = (n-q)/p;
    if(b>1 && f(b,n)==s){
      cout<<b<<endl;
      return 0;
    }
    p--;
  }

  cout<<-1<<endl;
  return 0;
}

File: work/arc060_d.py
# -*- coding: utf-8 -*-
import math
n = int(input())
s = int(input())

if s==n:
    print(n+1)
    exit()

def f(b,n):
    ret = 0
    while n>=b:
        ret += n%b
        n = n//b
    return ret+n

b = 2
while b*b<n:
    if f(b,n)==s:
        print(b)
        exit()
    b += 1

p = math.ceil(math.sqrt(n))+1
while p>0:
    q = s - p
    b = (n-q)//p
    if b>0 and f(b,n)==s:
        print(b)
        exit()
    p -= 1
print(-1)
